### DISCUSSION TITLE
Babel for dotnet maui.

### DISCUSSION THREAD

**Discussion by stefandb1995** [dotnet/maui#22572] [Q&A] [answered]
Hi!

I have a dotnet maui application that I want to obfuscate with Babel. I've got 2 projects that I want to merge (using Merge Assemblies). I found an old post on Xamarin.Android (https://github.com/xamarin/xamarin-android/issues/4202) that helped with the timings. In the bin folder of my main project I can see the merged .dll. Decompiling it does show that they are merged and obfuscated, but when I force a crash on the app and write the stacktrace to logcat I can read all of the code. By all accounts the merged dll isn't used in the apk, but I'm not sure what exactly I need to change.

Some code (so that we don't need to dig through the old post):

`
        <_BeforeLinkAssemblies>
            $(_BeforeLinkAssemblies);
            Obfuscate;
        </_BeforeLinkAssemblies>

	<Target Name="Obfuscate">
		<ItemGroup>
			<MainAndroidAssembly Remove="@(MainAndroidAssembly)" />
			<MainAndroidAssembly Include="@(ResolvedAssemblies)" Condition="'%(Filename)' == '$(AssemblyName)'" KeepMetadata="none" />
		</ItemGroup>

		<ItemGroup>
			<MergeAssemblies Remove="@(MergeAssemblies)" />
			<MergeAssemblies Include="@(ResolvedAssemblies)" Condition="'%(Filename)' == 'MergedAssembly1' Or '%(Filename)' == 'MergedAssembly2' Or '%(Filename)' == 'MergedAssembly3'" KeepMetadata="none" />
		</ItemGroup>

`

Then we have the obfuscation code. This should really matter, as there are other ways to merge assemblies.

`		<ItemGroup>
			<ResolvedAssemblies Remove="@(MergeAssemblies)" />
			<ResolvedUserAssemblies Remove="@(MergeAssemblies)" />
		</ItemGroup>
`

Any help would be appreciated!

**Comment by stefandb1995** [comment, marked as answer]
@bravekit I see you have a repo trying to figure this out as well?

**Reply by bravekit** [reply]
Hi,
yes, I was trying to figure out how to merge assemblies with Babel for my MAUI project. Eventually I decided to merge the assemblies just in my solution, because the "merge" function is available only with Babel Enterprise version (I have a trial licence now, but later I'm going to use a standard Babel licence). 
Anyway, I contacted to Babel's support, and Alberto sent me this:

The publish phase may vary depending on the target framework and platform. You might find the following task useful to include after ComputeFilesToPublish:
```
  <Target Name="UpdateBabelFilesToPublish"

          Condition="'$(BabelPublishEnabled)' == 'true'"

          DependsOnTargets="ObfuscateSettings"

          AfterTargets="ComputeFilesToPublish">

 

    <PropertyGroup>

      <MergedAssembliesRegEx>@(MergeAssembly->'%(FileName)(%(Extension)|.pdb|.xml)', '|')</MergedAssembliesRegEx>

      <EmbedAssembliesRegEx>@(EmbedAssembly->'%(FileName)(%(Extension)|.pdb|.xml)', '|')</EmbedAssembliesRegEx>

    </PropertyGroup>

 

    <!-- Remove merged and embedded assemblies from file to publish -->

    <ItemGroup>

      <ResolvedFileToPublish Remove="@(ResolvedFileToPublish->'%(Identity)')" Condition="('$(EmbedAssembliesRegEx)' != '') And $([System.Text.RegularExpressions.Regex]::IsMatch('%(FileName)%(Extension)', '$(EmbedAssembliesRegEx)'))" />

      <ResolvedFileToPublish Remove="@(ResolvedFileToPublish->'%(Identity)')" Condition="('$(MergedAssembliesRegEx)' != '') And $([System.Text.RegularExpressions.Regex]::IsMatch('%(FileName)%(Extension)', '$(MergedAssembliesRegEx)'))" />

    </ItemGroup>

    <Message Text="Babel publish files: @(ResolvedFileToPublish)" Importance="normal" />

  </Target>
```
I haven't tried it yet...

**Reply by stefandb1995** [reply]
Doesn't look like this timing step works for MAUI. It crashes the app when trying to use ComputeFilesToPublish.

**Reply by stefandb1995** [reply]
For anyone who is looking for this, it seems MAUI/dotnet android has fixed some timing (at least for our case) with the `<_BeforeLinkAssemblies>` timing in dotnet 9. So the above code is now working for us!

**Reply by blmiles** [reply]
Hello,

Just so you're aware of Babel limitations...

I have the Enterprise license. I can obfuscate and package Android and iOS versions of my app. Works very well.
BUT
When I tried packaging the Maui .msix for windows distribution, the Enterprise version won't do it.
Licensing issue.
They make you buy the Ultimate version. An additional 1250 euros or similar in US$

Consequently I've held off upgrading to Ultimate, although, if your company will spend the $, then you'd have no prob.
Short story: If you want to distribute into the Windows ecosystem, you'd need Ultimate to obfuscate and package.

**Reply by stefandb1995** [reply]
Luckily we only do Android and iOS, but thanks for the heads-up for anyone else dealing with this!
